src/components/tl/ucp/gather/gather_knomial.c

```diff
@@ -0,0 +1,171 @@
+            if (pos == 0) {
+                for (i = 1; i < radix; i++) {
+                    vpeer   = vrank + i * task->gather_kn.dist;
+                    msg_nbr = ucc_min(task->gather_kn.dist, team_size - vpeer);
+                    if (vpeer >= size) {
+                        break;
+                    } else if (vrank != 0) {
+                        msg_size       = data_size * msg_nbr;
+                        scratch_offset = PTR_OFFSET(
+                            scratch_offset, data_size * task->gather_kn.dist);
+                        peer = (vpeer + root) % size;
+                        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(scratch_offset,
+                                                         msg_size, mtype, peer,
+                                                         team, task),
+                                      task, out);
+                    } else { //The root is a particular case because it must aggregate the data sorted by ranks
+                        peer           = (vpeer + root) % size;
+                        scratch_offset = PTR_OFFSET(task->gather_kn.scratch,
+                                                    data_size * peer);
+                        if (msg_nbr <=
+                            team_size -
+                                peer) { // check if received data correspond to contiguous ranks
+                            msg_size = data_size * msg_nbr;
+                            UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(scratch_offset,
+                                                             msg_size, mtype,
+                                                             peer, team, task),
+                                          task, out);
+                        } else { // in this case, data must be split in two at the destination buffer
+                            msg_size = data_size * (team_size - peer);
+                            UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(scratch_offset,
+                                                             msg_size, mtype,
+                                                             peer, team, task),
+                                          task, out);
+
+                            msg_size =
+                                data_size * (msg_nbr - (team_size - peer));
+                            UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(
+                                              task->gather_kn.scratch, msg_size,
+                                              mtype, peer, team, task),
+                                          task, out);
+                        }
+                    }
+                }
+
+                if (task->gather_kn.dist == 1) { //check if first passage
+                    msg_size = data_size;
+                    if (rank != root) {
+                        ucc_mc_memcpy(task->gather_kn.scratch,
+                                      args->src.info.buffer, msg_size,
```

Comment:
alignment looks weird - is this what "git-clang format" produced ?

Reply:
I wrote it like this to avoid exceeding Col 80. I ran git-clang format afterward, I don't know if it changed this part. How should the alignment be?